src/verifier/snp/mod.rs

```diff
@@ -0,0 +1,141 @@
+use anyhow::{anyhow, Context, Result};
+extern crate serde;
+use self::serde::{Deserialize, Serialize};
+use super::*;
+use async_trait::async_trait;
+use serde_json::json;
+use sev::firmware::guest::types::AttestationReport;
+use sev::firmware::host::types::{CertTable, SnpCertType};
+
+use openssl::{
+    ec::EcKey,
+    ecdsa,
+    pkey::{PKey, Public},
+    x509,
+};
+use sha2::{Digest, Sha384};
+
+#[derive(Serialize, Deserialize)]
+struct SnpEvidence {
+    attestation_report: AttestationReport,
+    cert_chain: CertTable,
+}
+
+#[derive(Debug, Default)]
+pub struct Snp {}
+
+#[async_trait]
+impl Verifier for Snp {
+    async fn evaluate(
+        &self,
+        nonce: String,
+        attestation: &Attestation,
+    ) -> Result<TeeEvidenceParsedClaim> {
+        let tee_evidence = serde_json::from_str::<SnpEvidence>(&attestation.tee_evidence)
+            .context("Deserialize Quote failed.")?;
+
+        verify_report_signature(&tee_evidence)?;
+
+        let report = tee_evidence.attestation_report;
+        if report.version != 2 {
+            return Err(anyhow!("Unexpected report version"));
+        }
+
+        if report.vmpl != 0 {
+            return Err(anyhow!("VMPL Check Failed"));
+        }
+
+        let expected_report_data = calculate_expected_report_data(&nonce, &attestation.tee_pubkey);
+        if report.report_data != expected_report_data {
+            return Err(anyhow!("Report Data Mismatch"));
+        }
+
+        parse_tee_evidence(&report)
+    }
+}
+
+fn verify_report_signature(evidence: &SnpEvidence) -> Result<()> {
+    // check cert chain
+    let vcek = verify_cert_chain(&evidence.cert_chain)?;
+
+    // verify report signature
+    let sig = ecdsa::EcdsaSig::try_from(&evidence.attestation_report.signature)?;
+    let data = &bincode::serialize(&evidence.attestation_report)?[..=0x29];
+
+    sig.verify(&data, EcKey::try_from(vcek.public_key()?)?.as_ref())
+        .map_err(|_| anyhow!("Signature validation failed."))?;
+
+    Ok(())
+}
+
+fn verify_cert_chain(cert_chain: &CertTable) -> Result<x509::X509> {
+    // add support for genoa
+    let certs = x509::X509::stack_from_pem(include_bytes!("milan_ark_ask.pem"))?;
+    
+    // make this nicer
+    let ask = &certs[0];
+    let ark = &certs[1];
+
+    // check for revocation?
+
+    //let raw_vcek = cert_chain.entries.iter().find(|c| c.cert_type == SnpCertType::VCEK)
+    let raw_vcek = cert_chain
+        .entries
+        .iter()
+        .find(|c| c.guid() == "63da758d-e664-4564-adc5-f4b93be8accd")
```

Comment:
I don't believe the VCEK is part of the certificate chain you are referencing above. So this would almost always fail.

Reply:
I think we're expecting to VCEK to be provisioned at operator deployment time and to be part of he cert table in the extended report. Does that seem right?

Reply:
Ah, `certs` vs `cert_chain` my mistake.

Reply:
I'm going to clean up the parsing of the certs file and that should make it a bit more clear.

Reply:
also: extract the guid string to a const and add a comment referencing the GHCB spec section.

Reply:
I'm going to ditch this guid once he cert struct has the cert type as public.

Reply:
@fitzthum as of `1.1.0` you may now ditch this.